### PR TITLE
Add auto wrapping for `DDPFullyShardedNativeStrategy`

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

The PR resolves the note here: https://pytorch-lightning.readthedocs.io/en/latest/advanced/model_parallel.html?highlight=deepspeed#pytorch-fully-sharded-training

now users don't have to use `wrap` every time to shard the layers if they are switching to fully sharded.
I will do this to fairscale in a follow-up.

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda

### CODE REVIEW

File: docs/source-pytorch/advanced/model_parallel.rst [hunk 1]
```diff
@@ -212,14 +212,43 @@ PyTorch Fully Sharded Training
...
+It was introduced in their `v1.11.0 release <https://pytorch.org/blog/introducing-pytorch-fully-sharded-data-parallel-api/>`_ but it is recommended to use it with PyTorch v1.12 or more and that's what
+Lightning supports. The API is pretty similar to that of FairScale.
 
-.. note::
-    Currently Fully Sharded Training relies on the user to wrap the model with Fully Sharded within the ``LightningModule``.
-    This means you must create a single model that is treated as a ``torch.nn.Module`` within the ``LightningModule``.
-    This is a limitation of Fully Sharded Training that will be resolved in the future.
 
-To activate parameter sharding, you must wrap your model using the``wrap`` function. Internally in Lightning, we enable a context manager around the ``configure_sharded_model`` function to make sure the ``wrap`` parameters are passed correctly.
+Auto Wrapping
```
Comment (awaelchli) [inline docs/source-pytorch/advanced/model_parallel.rst:219]: I think we should show the opposite:

Auto:
Brief description that when using fsdp native, user does not have to wrap layers manually. No example code needed, because there is nothing the user needs to do in the model

Manual:
Here describe how and when to do the wrapping manually. Then show the example code.

File: src/pytorch_lightning/strategies/fully_sharded_native.py [hunk 2]
```diff
@@ -201,6 +206,26 @@ def _configure_launcher(self) -> None:
             self._launcher = _SubprocessScriptLauncher(self.cluster_environment, self.num_processes, self.num_nodes)
             self._rank_0_will_call_children_scripts = True
 
+    def _setup_model(self, model: torch.nn.Module) -> FullyShardedDataParallel:
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply src/pytorch_lightning/strategies/fully_sharded_native.py:209]: This is the signature of this method, so this looks correct.

File: src/pytorch_lightning/strategies/fully_sharded_native.py [hunk 3]
```diff
@@ -273,6 +310,22 @@ def reduce(
             tensor = sync_ddp_if_available(tensor, group, reduce_op=reduce_op)
         return tensor
 
+    def training_step(self, *args: Any, **kwargs: Any) -> STEP_OUTPUT:
```
Comment (awaelchli) [inline src/pytorch_lightning/strategies/fully_sharded_native.py:313]: Why aren't we using the precision context here? Please add a comment.

File: src/pytorch_lightning/strategies/fully_sharded_native.py [hunk 4]
```diff
@@ -215,9 +240,21 @@ def setup(self, trainer: "pl.Trainer") -> None:
         assert self.lightning_module is not None
         self.lightning_module._device = self.root_device
 
+        assert isinstance(self.model, pl.LightningModule)
+        self.model = _LightningModuleWrapperBase(self.model)
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply src/pytorch_lightning/strategies/fully_sharded_native.py:244]: This is because @rohitgr7 added the FSDP wrapper around the LM. When we call model.forward() below, we still need to redirect to training_step or whatever *_step internally, based on the stage, and this is what this `_LightningModuleWrapperBase` does.

**Review by awaelchli — CHANGES_REQUESTED** [review]

File: src/pytorch_lightning/strategies/fully_sharded_native.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (justusschock) [reply src/pytorch_lightning/strategies/fully_sharded_native.py:244]: Yes, but why do it different now than before?

File: src/pytorch_lightning/strategies/fully_sharded_native.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (justusschock) [reply src/pytorch_lightning/strategies/fully_sharded_native.py:209]: Okay, but then I suggest we use the model passed there instead of relying on `self.lightning_module`

[the remark this answers is not in the document]
Comment (awaelchli) [reply src/pytorch_lightning/strategies/fully_sharded_native.py:209]: Yes I agree. @rohitgr7 You can look how it's done in DDP.

File: src/pytorch_lightning/strategies/fully_sharded_native.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (awaelchli) [reply src/pytorch_lightning/strategies/fully_sharded_native.py:244]: I think because before, we didn't wrap the lightning module at all. Now that we do, we need to do the same as in DDP.

File: src/pytorch_lightning/strategies/fully_sharded_native.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rohitgr7) [reply src/pytorch_lightning/strategies/fully_sharded_native.py:209]: it's done like this in DDP too. The reason I didn't change it because it was raising mypy error because of different base class signature.

and something related to Lite is mentioned there:
```
# TODO (@awaelchli): standardize this across all plugins in Lightning and Lite. Related refactor: #7324
```
so I didn't change the signature.

File: src/pytorch_lightning/strategies/fully_sharded_native.py [hunk 5]
```diff
@@ -41,6 +42,9 @@
 from pytorch_lightning.utilities.optimizer import optimizers_to_device
 from pytorch_lightning.utilities.rank_zero import rank_zero_info
 from pytorch_lightning.utilities.seed import reset_seed
+from pytorch_lightning.utilities.types import STEP_OUTPUT
+
+from ..utilities.model_helpers import is_overridden
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply src/pytorch_lightning/strategies/fully_sharded_native.py:47]: ah! auto import 🤦

File: src/pytorch_lightning/strategies/fully_sharded_native.py [hunk 6]
```diff
@@ -215,9 +240,21 @@ def setup(self, trainer: "pl.Trainer") -> None:
...
 
+        assert isinstance(self.model, pl.LightningModule)
+        self.model = _LightningModuleWrapperBase(self.model)
+        if is_overridden("configure_sharded_model", self.lightning_module):
+            rank_zero_info(
+                "You have overridden `LightningModule.configure_sharded_model` hook. It will assume that all the layers"
+                " are already wrapped for sharding and won't wrap the entire model using `FullyShardedDataParallel`."
+            )
+        else:
+            self.model = self._setup_model(self.model)
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply src/pytorch_lightning/strategies/fully_sharded_native.py:251]: answered above!

File: src/pytorch_lightning/strategies/fully_sharded_native.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rohitgr7) [reply src/pytorch_lightning/strategies/fully_sharded_native.py:244]: so the input cast in case of precision happens inside the Fullyshardedwrapper, which internally calls the base modules forward method. I did this to ensure we have a forward method available that gets the casted input and then the lightningmodule methods are called accordingly. Previously I guess it wasn't done because the wrap was done by the user manually, so there's always a wrapper around each individual layer.

I think it might work without `_LightningModuleWrapperBase`, but just wanted to be consistent here and avoid any edge cases with casting.

File: src/pytorch_lightning/strategies/fully_sharded_native.py (hunk 3 above)
Comment (rohitgr7) [reply src/pytorch_lightning/strategies/fully_sharded_native.py:312]: because casting is done by FSDP. add a comment.

File: tests/tests_pytorch/strategies/test_ddp_fully_sharded_native.py [hunk 7]
```diff
@@ -140,18 +183,32 @@ def test_fully_sharded_native_strategy_checkpoint(tmpdir, precision):
 
 
 @RunIf(min_cuda_gpus=2, skip_windows=True, standalone=True, min_torch="1.12")
-def test_fully_sharded_native_strategy_checkpoint_multi_gpus(tmpdir):
+@pytest.mark.parametrize(
+    "model, strategy",
+    [
+        (TestFSDPModel(), "fsdp_native"),
+        (TestFSDPModelAutoWrapped(), DDPFullyShardedNativeStrategy),
```
Comment (awaelchli) [inline tests/tests_pytorch/strategies/test_ddp_fully_sharded_native.py:190]: Since you are adding new parameterizations here, please make sure that the increase in test time is not unreasonable.

**Review by awaelchli — APPROVED** [review]

File: tests/tests_pytorch/strategies/test_ddp_fully_sharded_native.py (hunk 7 above)
Comment (rohitgr7) [reply tests/tests_pytorch/strategies/test_ddp_fully_sharded_native.py:190]: it's not much locally. Just a few seconds.

File: src/pytorch_lightning/plugins/precision/fsdp_native_native_amp.py [hunk 8]
```diff
@@ -25,14 +25,13 @@
...
+    ShardedGradScaler = None  # type: ignore[misc,assignment]
 
 
 class FullyShardedNativeNativeMixedPrecisionPlugin(NativeMixedPrecisionPlugin):
     """Native AMP for Fully Sharded Native Training."""
 
-    def __init__(
-        self, precision: Union[str, int], device: str, scaler: Optional[torch.cuda.amp.GradScaler] = None
-    ) -> None:
+    def __init__(self, precision: Union[str, int], device: str, scaler: Optional[ShardedGradScaler] = None) -> None:
```
Comment (carmocca) [inline src/pytorch_lightning/plugins/precision/fsdp_native_native_amp.py:34]: The `ShardedGradScaler` subclasses the `cuda.amp.GradScaler`. Did you change this because mypy was complaining?

Does the code require this specific subclass?

If yes, then the annotation for `src/pytorch_lightning/plugins/precision/sharded_native_amp.py` should also be updated.

File: src/pytorch_lightning/strategies/fully_sharded_native.py [hunk 9]
```diff
@@ -201,6 +205,26 @@ def _configure_launcher(self) -> None:
...
             self._rank_0_will_call_children_scripts = True
 
+    def _setup_model(self, model: torch.nn.Module) -> FullyShardedDataParallel:
+        """Wraps the model into a
+        :class:`~torch.distributed.fsdp.fully_sharded_data_parallel.FullyShardedDataParallel` module."""
+        # If model is already wrapped, we need to avoid sending the `auto_wrap_policy`
+        assert self.lightning_module is not None
+        if any(isinstance(mod, FullyShardedDataParallel) for _, mod in self.lightning_module.named_modules()):
+            if "auto_wrap_policy" in self.kwargs:
+                self.kwargs.pop("auto_wrap_policy")
```
Comment (carmocca) [inline src/pytorch_lightning/strategies/fully_sharded_native.py:215]: - Merge the ifs
- Names are unused
- we can safely del since you just checked it's inside

```suggestion
        if any(isinstance(mod, FullyShardedDataParallel) for mod in self.lightning_module.modules()) and "auto_wrap_policy" in self.kwargs:
            del kwargs["auto_wrap_policy"]
```

File: src/pytorch_lightning/strategies/fully_sharded_native.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (carmocca) [reply src/pytorch_lightning/strategies/fully_sharded_native.py:208]: > it's done like this in DDP too

I don't get this comment, in `DDPStrategy._setup_model` the input argument is what gets wrapped into `DistributedDataParallel`

File: src/pytorch_lightning/strategies/fully_sharded_native.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (carmocca) [reply src/pytorch_lightning/strategies/fully_sharded_native.py:243]: That would explain why we wrap with `_LightningModuleWrapperBase`, but why not use `self.lightning_module`?

File: src/pytorch_lightning/strategies/fully_sharded_native.py [hunk 10]
```diff
@@ -215,9 +239,21 @@ def setup(self, trainer: "pl.Trainer") -> None:
         assert self.lightning_module is not None
         self.lightning_module._device = self.root_device
 
+        assert isinstance(self.model, pl.LightningModule)
+        self.model = _LightningModuleWrapperBase(self.model)
+        if is_overridden("configure_sharded_model", self.lightning_module):
```
Comment (carmocca) [inline src/pytorch_lightning/strategies/fully_sharded_native.py:244]: I would merge this with the checks for popping "auto_wrap_policy" in `_setup_model`. Either by pulling those to here or moving this into `_setup_model`

File: src/pytorch_lightning/strategies/fully_sharded_native.py [hunk 11]
```diff
@@ -215,9 +239,21 @@ def setup(self, trainer: "pl.Trainer") -> None:
...
+                "You have overridden `LightningModule.configure_sharded_model` hook. It will assume that all the layers"
+                " are already wrapped for sharding and won't wrap the entire model using `FullyShardedDataParallel`."
+            )
+        else:
+            self.model = self._setup_model(self.model)
         self.barrier()
-        self.setup_optimizers(trainer)
-        optimizers_to_device(self.optimizers, self.root_device)
+
+        if trainer.state.fn == TrainerFn.FITTING:
```
Comment (carmocca) [inline src/pytorch_lightning/strategies/fully_sharded_native.py:253]: Was this a bug? Should it be separated for a bug-fix release?

File: src/pytorch_lightning/strategies/fully_sharded_native.py [hunk 12]
```diff
@@ -273,6 +309,24 @@ def reduce(
             tensor = sync_ddp_if_available(tensor, group, reduce_op=reduce_op)
         return tensor
 
+    def training_step(self, *args: Any, **kwargs: Any) -> STEP_OUTPUT:
+        # we don't need precision context since casting is done by FSDP
+        # read `mixed_precision``` docstring here: https://pytorch.org/docs/stable/fsdp.html
```
Comment (carmocca) [inline src/pytorch_lightning/strategies/fully_sharded_native.py:314]:
```suggestion
        # read `mixed_precision` docstring here: https://pytorch.org/docs/stable/fsdp.html
```

Also, you can :paramref: the argument of the FSDP class in PyTorch docs

File: src/pytorch_lightning/strategies/fully_sharded_native.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rohitgr7) [reply src/pytorch_lightning/strategies/fully_sharded_native.py:208]: yes, and here also the input argument is wrapped
```py
return FullyShardedDataParallel(
            module=model,
```

File: src/pytorch_lightning/plugins/precision/fsdp_native_native_amp.py (hunk 8 above)
Comment (rohitgr7) [reply src/pytorch_lightning/plugins/precision/fsdp_native_native_amp.py:34]: no, mypy wasn't failing. I just wanted to be explicit to make sure users passes the `ShardedGradScaler` with shared strategies. If you agree on this, will update `sharded_native_amp` too else revert this.

File: src/pytorch_lightning/strategies/fully_sharded_native.py (hunk 11 above)
Comment (rohitgr7) [reply src/pytorch_lightning/strategies/fully_sharded_native.py:253]: yes, it was. will send a fix for this and remove it from here.

File: src/pytorch_lightning/strategies/fully_sharded_native.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rohitgr7) [reply src/pytorch_lightning/strategies/fully_sharded_native.py:243]: because if we do that, then each layer will act as root w.r.t FSDP and will create this issue:
https://github.com/pytorch/pytorch/blob/67ece03c8cd632cce9523cd96efde6f2d1cc8121/torch/distributed/fsdp/fully_sharded_data_parallel.py#L1464-L1475

also, the tests fail if I use `lightning_module` and use the `*_step` methods directly. The inputs should be passed from the wrapped module first to ensure each sub-layer should not act as root.

File: docs/source-pytorch/advanced/model_parallel.rst [hunk 13]
```diff
@@ -212,14 +212,28 @@ PyTorch Fully Sharded Training
...
+"""""""""""""
+Model layers should be wrapped in FSDP in a nested way to save peak memory and enable communication and computation overlapping. The
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply docs/source-pytorch/advanced/model_parallel.rst:221]: I think it just means a normal nested sequential form.
```
module(
    layer 1,
    layer 2,
    sequence(
        layer3,
        layer4,
    )
```
took it from herE: https://pytorch.org/blog/introducing-pytorch-fully-sharded-data-parallel-api/#auto-wrapping

File: src/pytorch_lightning/strategies/fully_sharded_native.py (hunk 10 above)
Comment (rohitgr7) [reply src/pytorch_lightning/strategies/fully_sharded_native.py:244]: `auto_wrap_policy` is part of `kwargs` used within the sharding context manager, so we shouldn't do this.

File: src/pytorch_lightning/strategies/fully_sharded_native.py (hunk 12 above)
Comment (rohitgr7) [reply src/pytorch_lightning/strategies/fully_sharded_native.py:314]: it's a comment. Will that be useful?

File: src/pytorch_lightning/plugins/precision/fsdp_native_native_amp.py (hunk 8 above)
Comment (carmocca) [reply src/pytorch_lightning/plugins/precision/fsdp_native_native_amp.py:34]: It's a bit restrictive but I'm fine with doing it for clarity. 

> will update sharded_native_amp

:+1:

File: docs/source-pytorch/advanced/model_parallel.rst (hunk 13 above)
[the remark this answers is not in the document]
Comment (carmocca) [reply docs/source-pytorch/advanced/model_parallel.rst:221]: Okay... should probably reference that link

**Review by carmocca — APPROVED** [review]

**Review by justusschock — APPROVED** [review]
